pkg/api/shared/status.go

```diff
@@ -0,0 +1,71 @@
+package shared
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients"
+	"github.com/solo-io/solo-kit/pkg/api/v1/resources"
+	"github.com/solo-io/solo-kit/pkg/errors"
+)
+
+// ApplyStatus is used by clients that don't support patch updates to resource statuses (e.g. consul, files, in-memory)
+func ApplyStatus(rc clients.ResourceClient, statusClient resources.StatusClient, inputResource resources.InputResource, opts clients.ApplyStatusOpts) (resources.Resource, error) {
+	name := inputResource.GetMetadata().GetName()
+	namespace := inputResource.GetMetadata().GetNamespace()
+	res, err := rc.Read(namespace, name, clients.ReadOpts{
+		Ctx:     opts.Ctx,
+		Cluster: opts.Cluster,
+	})
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading before applying status")
+	}
+
+	inputRes, ok := res.(resources.InputResource)
+	if !ok {
+		return nil, errors.Errorf("error converting resource of type %T to input resource to apply status", res)
+	}
+
+	statusClient.SetStatus(inputRes, statusClient.GetStatus(inputResource))
+	updatedRes, err := rc.Write(inputRes, clients.WriteOpts{
+		Ctx:               opts.Ctx,
+		OverwriteExisting: true,
+	})
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "error writing to apply status")
+	}
+	return updatedRes, nil
+}
+
+// GetJsonPatchData returns the json patch data for the input resource.
+// Prefer using json patch for single api call status updates when supported (e.g. k8s) to avoid ratelimiting
+// to the k8s apiserver (e.g. https://github.com/solo-io/gloo/blob/a083522af0a4ce22f4d2adf3a02470f782d5a865/projects/gloo/api/v1/settings.proto#L337-L350)
+func GetJsonPatchData(inputResource resources.InputResource) ([]byte, error) {
+	namespacedStatuses := inputResource.GetNamespacedStatuses().GetStatuses()
+	if len(namespacedStatuses) != 1 {
+		// we only expect our namespace to report here; we don't want to blow away statuses from other reporters
+		return nil, errors.Errorf("unexpected number of namespaces in input resource: %v", len(inputResource.GetNamespacedStatuses().GetStatuses()))
+	}
+	ns := ""
+	for loopNs := range inputResource.GetNamespacedStatuses().GetStatuses() {
+		ns = loopNs
+	}
+	status := inputResource.GetNamespacedStatuses().GetStatuses()[ns]
+
+	buf := &bytes.Buffer{}
+	var marshaller jsonpb.Marshaler
+	marshaller.EnumsAsInts = false  // prefer jsonpb over encoding/json marshaller since it renders enum as string not int (i.e., state is human-readable)
```

Comment:
Currently we write out ints for the states. If we want to switch to writing strings we should  call it out in the changelog. I assume that this change won't affect any tooling around statuses because the unmarshaller will handle it the same regardless.

Reply:
happy to call that out in the changelog. yeah we currently use the jsonpb to unmarshal so it can handle either format, so im just opting for better readability